### PR TITLE
Model Animation Updates

### PR DESCRIPTION
**Pull request by mdiller**
This pull request implements the following:
- Better access to frames from Animation.cs
- Adds animations to models exported as GLTF
- Adds support for exporting to GLB instead of GLTF as an option in the GUI
- Adds some play controls to the animations in the model viewer for the GUI

Screenshot of new GLB option:
image

GIF showing new animation controls:
animationcontrols

GIF showing animations exported properly into GLTF and then played via blender:
animationsworking

If one of these features isn't wanted, they're fairly separated in the commits I made so I can revert a commit or 2 if need be.

Ref: #232

### CODE REVIEW

File: ValveResourceFormat/IO/GltfModelExporter.cs [hunk 1]
```diff
@@ -68,6 +69,42 @@ void AddMeshNode(string name, VMesh mesh, Skeleton skeleton)
...
+
+                    // Add animations
+                    var animations = GetAllAnimations(model);
+                    foreach (var animation in animations)
+                    {
+                        var exportedAnimation = exportedModel.CreateAnimation(animation.Name);
+                        var rotationDict = new Dictionary<string, Dictionary<float, Quaternion>>();
+                        var translationDict = new Dictionary<string, Dictionary<float, Vector3>>();
+
+                        var time = (float)0;
```
Comment (xPaw) [inline ValveResourceFormat/IO/GltfModelExporter.cs:81]:
```suggestion
                        var time = 0f;
```

File: ValveResourceFormat/IO/GltfModelExporter.cs [hunk 2]
```diff
@@ -86,6 +123,8 @@ void AddMeshNode(string name, VMesh mesh, Skeleton skeleton)
                 AddMeshNode(meshes[i].Name, meshes[i].Mesh, model.GetSkeleton(i));
             }
 
+
+
```
Comment (xPaw) [inline ValveResourceFormat/IO/GltfModelExporter.cs:127]: there's a few pointless lines that got added in this file

File: ValveResourceFormat/IO/GltfModelExporter.cs [hunk 3]
```diff
@@ -474,6 +514,40 @@ private Material GenerateGLTFMaterialFromRenderMaterial(VMaterial renderMaterial
...
+            var animGroupPaths = model.GetReferencedAnimationGroupNames();
+            var animations = model.GetEmbeddedAnimations().ToList();
+
+            // Load animations from referenced animation groups
+            foreach (var animGroupPath in animGroupPaths)
+            {
+                var animGroup = FileLoader.LoadFile(animGroupPath + "_c");
+                if (animGroup != default)
+                {
+                    var data = animGroup.DataBlock is ResourceTypes.NTRO ntro
```
Comment (xPaw) [inline ValveResourceFormat/IO/GltfModelExporter.cs:528]: I believe we have a helper function that abstracts NTRO/BinaryKV3 getter.

File: ValveResourceFormat/Resource/ResourceTypes/ModelAnimation/Animation.cs [hunk 4]
```diff
@@ -12,10 +12,8 @@ public class Animation
     {
         public string Name { get; private set; }
         public float Fps { get; private set; }
-
-        private long FrameCount;
-
-        private Frame[] Frames;
+        public int FrameCount { get; private set; }
+        public IReadOnlyList<Frame> Frames { get; set; }
```
Comment (xPaw) [inline ValveResourceFormat/Resource/ResourceTypes/ModelAnimation/Animation.cs:16]: private set?

File: ValveResourceFormat/Resource/ResourceTypes/ModelAnimation/Animation.cs [hunk 5]
```diff
@@ -166,14 +164,15 @@ private Frame GetTransformsAtTime(float time)
                 : pDataObject as IKeyValueCollection;
             var frameBlockArray = pData.GetArray("m_frameblockArray");
 
-            FrameCount = pData.GetIntegerProperty("m_nFrames");
-            Frames = new Frame[FrameCount];
+            FrameCount = (int)pData.GetIntegerProperty("m_nFrames");
+            var frameArray = new Frame[FrameCount];
+            Frames = frameArray;
```
Comment (xPaw) [inline ValveResourceFormat/Resource/ResourceTypes/ModelAnimation/Animation.cs:169]: For the sake of clarity, I would have assigned this after the loop.

File: ValveResourceFormat/IO/GltfModelExporter.cs (hunk 1 above)
Comment (mdiller) [reply ValveResourceFormat/IO/GltfModelExporter.cs]: updated as suggested

File: ValveResourceFormat/IO/GltfModelExporter.cs (hunk 2 above)
Comment (mdiller) [reply ValveResourceFormat/IO/GltfModelExporter.cs]: fixed extra lines in this file

File: ValveResourceFormat/Resource/ResourceTypes/ModelAnimation/Animation.cs (hunk 4 above)
Comment (mdiller) [reply ValveResourceFormat/Resource/ResourceTypes/ModelAnimation/Animation.cs]: whoops, fixed

File: ValveResourceFormat/Resource/ResourceTypes/ModelAnimation/Animation.cs (hunk 5 above)
Comment (mdiller) [reply ValveResourceFormat/Resource/ResourceTypes/ModelAnimation/Animation.cs]: gotcha. I've made this change now

File: ValveResourceFormat/IO/GltfModelExporter.cs (hunk 3 above)
Comment (mdiller) [reply ValveResourceFormat/IO/GltfModelExporter.cs:525]: I can't find one? I took this implementation from `AnimationGroupLoader.cs`.

Comment (mdiller) [reply ValveResourceFormat/IO/GltfModelExporter.cs]: Fixed as suggested via discord

**Comment by mdiller** [comment]
@xPaw I ran the decompiler with `Decompiler.exe -i "...\dota 2 beta\game\dota\pak01_dir.vpk" --threads
32 --vpk_extensions="vmdl_c" --output "...\test_out" --vpk_decompile` as requested, and these are the exceptions that it spat out. 
[exceptions.txt](https://github.com/SteamDatabase/ValveResourceFormat/files/5563718/exceptions.txt)
Looks like almost all of them are variations on:
```
Exception: SharpGLTF.Validation.DataException: Accessor[12] JOINTS_0[118]: Is out of bounds
```

And one is:
```
File: models\props_frostivus\frostivus_ancient\frostivus_ancient_destruction.vmdl_c
Exception: SharpGLTF.Validation.SchemaException: MeshPrimitive[0] JOINTS_0: 13411 must be equal to 26822.
```

These don't mention animations so I'm guessing these exceptions already existed?

**Comment by xPaw** [comment]
These have open issues already.

**Review by xPaw — APPROVED** [review]

**Comment by mdiller** [comment]
squashed commits as requested

**Review by Perryvw — APPROVED** [review]

**Pull request merged**